src/Route.php

```diff
@@ -169,17 +183,58 @@ public function defaults(array $defaults): self
     }
 
     /**
-     * @param callable|MiddlewareInterface $middleware
-     * @return MiddlewareInterface
+     * @param MiddlewareInterface|callable|string|array $middleware
      */
-    private function prepareMiddleware($middleware): MiddlewareInterface
+    private function validateMiddleware($middleware): void
     {
+        if (
+            is_string($middleware) && is_subclass_of($middleware, MiddlewareInterface::class)
+        ) {
+            if ($this->container === null) {
+                throw new InvalidArgumentException('Route container must not be null for lazy loaded middleware.');
+            }
+            return;
+        }
+
+        if (
+            is_array($middleware) && count($middleware) === 2
+            && isset($middleware[0], $middleware[1])
+            && is_string($middleware[1]) && is_string($middleware[0]) && class_exists($middleware[0])
+        ) {
+            if ($this->container === null) {
+                throw new InvalidArgumentException('Route container must not be null for handler action.');
+            }
+            return;
+        }
+
         if (is_callable($middleware)) {
-            $middleware = new Callback($middleware);
+            if ($this->container === null) {
+                throw new InvalidArgumentException('Route container must not be null for callable.');
+            }
+            return;
         }
 
         if (!$middleware instanceof MiddlewareInterface) {
-            throw new InvalidArgumentException('Parameter should be either a PSR middleware or a callable.');
+            throw new InvalidArgumentException('Parameter should be either PSR middleware instance, PSR middleware class name, handler action or a callable.');
+        }
+    }
+
+    /**
+     * @param MiddlewareInterface|callable|string|array $middleware
+     * @return MiddlewareInterface|string|array
+     */
+    private function prepareMiddleware($middleware)
+    {
+        if (is_string($middleware)) {
+            return $this->container->get($middleware);
+        }
+
+        if (is_array($middleware)) {
```

Comment:
A callable can be an array. See my last commit with added test and fix.

Reply:
@yiiliveext `['ClassName', 'methodName']` is a callable too

Reply:
> @yiiliveext `['ClassName', 'methodName']` is a callable too

That's what I said.